sdk/src/vote_program.rs

```diff
@@ -186,25 +186,64 @@ impl VoteState {
     }
 }
 
-// TODO: Deprecate the RegisterAccount instruction and its awkward delegation
-// semantics.
-pub fn register(keyed_accounts: &mut [KeyedAccount]) -> Result<(), ProgramError> {
+pub fn delegate_account(
+    keyed_accounts: &mut [KeyedAccount],
+    node_id: Pubkey,
+) -> Result<(), ProgramError> {
+    if !check_id(&keyed_accounts[0].account.owner) {
+        error!("account[0] is not assigned to the VOTE_PROGRAM");
+        Err(ProgramError::InvalidArgument)?;
+    }
+
+    if keyed_accounts[0].signer_key().is_none() {
+        error!("account[0] should sign the transaction");
+        Err(ProgramError::InvalidArgument)?;
+    }
+
+    let vote_state = VoteState::deserialize(&keyed_accounts[0].account.userdata);
+    if let Ok(mut vote_state) = vote_state {
+        vote_state.delegate_id = node_id;
+        if vote_state.staker_id == Pubkey::default() {
+            vote_state.staker_id = node_id
+        }
+        vote_state.serialize(&mut keyed_accounts[0].account.userdata)?;
+    } else {
+        error!("account[1] does not have enough space assigned");
```

Comment:
That error should be `account[0]`. And pretty clearly a missing test here,  as this UserdataTooSmall error should apply to the `serialize()` above,  not this `deserialize()`.